proposals/lotus-blockstore-interface.md

```diff
@@ -0,0 +1,44 @@
+# Extended Blockstore Interface
+
+Authors: vyzo
+
+## What is the problem this project solves?
+
+We currently have a limited "dumb" blockstore interface, which does not allow us to pass context options so as to inform the blockstore about certain properties of objects being stored.
+
+This is particularly important in the context of the splitstore, where we have the following problems:
+- We lack the ability to inform the splitstore of the actual epoch an object belongs, which requires us to "guess" by estimating based int he current tipset and wall clock (see [#6474](https://github.com/filecoin-project/lotus/pull/6474)). This is problematic for many reasons, as an object might be misassigned its epoch (and its completely broken during catch up sync in current master).
+- We lack the ability to inform the splitstore when an access is initiated from the network. In conjunction with the transactional gc logic in [#6474](https://github.com/filecoin-project/lotus/pull/6474) this would result in network requests keeping objects that would otherwise be pruged from the hotstore live.
+This is both a performance/gc issue (the hotstore is larger than it needs to be) but it is also a potential attack vector for nodes that need to run with fixed hardware (e.g. network boosters).
+- We lack the ability to inform the splitstore that certain objects should be reified from cold storage to hot storage. For instance, a miner doing tipset state computation in an older tipset may want to keep this in the hot storage. It also means that misses result in permanent banishment in the coldstore and makes it hard to gc by simply removing the coldstore.
+
+## Impact
+
+This will make the splitstore a much more reliable component, working towards making it the default blockstore in lotus.
+
+## The idea
+
+Extend the blockstore interface so as to accept an `Options` struct, which specifies the necessary context constraints.
+
+Methods would look like this:
+```
+type Blockstore interface {
+  ...
+  Get(c cid.Cid, opts Options) (blocks.Block, error)
+  ...
+}
+```
+
+We may want to name the new option accepting methods with an `WithOptions` suffix if we want to keep backwards compatibility in the interface.
+
+Note: The reason for using a flat `Options` struct, as opposed to a `context.Context` object or functional options is purely performance. The latter options can have quite a bit of overhead, which we want to eliminate.
```

Comment:
We want to put contexts everywhere anyway for the purposes of debugging/tracing as well as general extensibility https://github.com/ipfs/go-ipfs/issues/6803.

Is the performance of `context.WithValue` really that slow? If so then I could see options making sense here.

Reply:
it can be quite bad in my experience, especially with multiple options and compound contexts.

Reply:
`context.WithValue` is also not typed.

Reply:
There's no way we can both make `Options` "typed" _and_ support passing things like epochs (which I don't think we should do anyways).

Reply:
If we're breaking the interfaces anyway any reason not to add in a context? Seems like it'd be helpful if we land the changes to the blockstore + datastore linked above which we're hoping to do.